En este Notebook se ejecutan los comandos necesarios para entrenar y evaluar los 4 modelos predictivos (Logistic regression, Standard LSTM, Channel-wise LSTM y Channel-wise Gru) diseñados para la tarea 'Predicción de LOS >3 días'. 

Nota: Para obtener los archivos en el los3.zip primero ejecuta los siguientes comandos (para más detalles véase 'pasos_de_comandos_LOS3days.txt'): 

1. python3 -m mimic3benchmark.scripts.extract_subjects {PATH TO MIMIC-III CSVs} data/root/
2. python3 -m mimic3benchmark.scripts.validate_events data/root/
3. python3 -m mimic3benchmark.scripts.extract_episodes_from_subjects data/root/
4. python3 -m mimic3benchmark.scripts.split_train_and_test data/root/
5. python3 -m mimic3benchmark.scripts.create_losxdays data/root/ data/{task name}/ {un entero que indica días de estancia}
6. python3 -m mimic3models.create_normalizer_state --impute_strategy previous --n_samples -1 --output_dir . --start_time zero --store_masks --task los --timestep 1.0 --data {PATH TO data/{task name}}
7. python -m mimic3models.split_train_val {dataset-directory}

Observación: el notebook se ejecutó en Google Colab.

### Setups:

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [ ]:
!ls

gdrive	los3.zip  sample_data


In [ ]:
%cd /content

/content


In [ ]:
!ls

gdrive	los3.zip  sample_data


In [ ]:
%tensorflow_version 1.x

TensorFlow 1.x selected.


In [ ]:
import tensorflow 
print(tensorflow.__version__)

1.15.2


In [ ]:
#!pip3 install tensorflow==1.15.2

In [ ]:
!pip install keras==2.1.2

  Using cached https://files.pythonhosted.org/packages/68/89/58ee5f56a9c26957d97217db41780ebedca3154392cb903c3f8a08a52208/Keras-2.1.2-py2.py3-none-any.whl
ERROR: textgenrnn 1.4.1 has requirement keras>=2.1.5, but you'll have keras 2.1.2 which is incompatible.
  Found existing installation: Keras 2.4.3
    Uninstalling Keras-2.4.3:
      Successfully uninstalled Keras-2.4.3


In [ ]:
import keras 
print(keras.__version__)

2.1.2


Using TensorFlow backend.


In [ ]:
!unzip los3.zip

Se han truncado las últimas 5000 líneas del flujo de salida.
  inflating: los3days/train/31568_episode1_timeseries.csv  
  inflating: los3days/train/83557_episode1_timeseries.csv  
  inflating: los3days/train/17080_episode1_timeseries.csv  
  inflating: los3days/train/28700_episode1_timeseries.csv  
  inflating: los3days/train/2358_episode1_timeseries.csv  
  inflating: los3days/train/98748_episode1_timeseries.csv  
  inflating: los3days/train/15121_episode1_timeseries.csv  
  inflating: los3days/train/30552_episode4_timeseries.csv  
  inflating: los3days/train/21123_episode2_timeseries.csv  
  inflating: los3days/train/21123_episode1_timeseries.csv  
  inflating: los3days/train/8674_episode2_timeseries.csv  
  inflating: los3days/train/29502_episode2_timeseries.csv  
  inflating: los3days/train/2427_episode1_timeseries.csv  
  inflating: los3days/train/6440_episode3_timeseries.csv  
  inflating: los3days/train/6440_episode8_timeseries.csv  
  inflating: los3days/train/66043_episode1_t

In [ ]:
%cd /content/gdrive/MyDrive/MIMIC_benchmark

/content/gdrive/MyDrive/MIMIC_benchmark


In [ ]:
!ls

50epochsOutputLSTMMortality.txt
data
ihm_results_channelWise_epoch34.json
ihm_results_channelWise_gru_epoch20.json
ihm_results_evaluation_channelWise_epoch14.json
ihm_results_evaluation_logistic.json
ihm_results_lstm_epoch33.json
LICENSE
los3days_results_channelWise_lstm_epoch28.json
los3days_results_channelWise_lstm.json
los3days_results_logistic.json
los3days_results_lstm_epoch35.json
los3days_results_lstm.json
los7days_results_channelWise_lstm.json
los7days_results_logistic.json
los7days_results_lstm.json
los_results_channelWise.json
los_results_logistic.json
los_results_lstm.json
mimic3benchmark
mimic3models
README.md
requirements.txt
scripts_pipeline_forLOS3days.txt
scripts_pipeline_forLOS7days.txt
scripts_pipeline_forLOSCustomized.txt
scripts_pipeline.txt
statistics.md


## Logistic Regression:

In [ ]:
!python -um mimic3models.los3days.logistic.main --data /content/los3days --l2 --C 0.001 --output_dir mimic3models/los3days/logistic

Namespace(C=0.001, data='/content/los3days', features='all', l2=True, output_dir='mimic3models/los3days/logistic', period='all')
Reading data and extracting features ...
  train data shape = (24490, 714)
  validation data shape = (5356, 714)
  test data shape = (5282, 714)
Imputing missing values ...
Normalizing the data to have zero mean and unit variance ...
confusion matrix:
[[13208  1871]
 [ 4577  4834]]
accuracy = 0.7367088794708252
precision class 0 = 0.7426483035087585
precision class 1 = 0.7209545373916626
recall class 0 = 0.8759201765060425
recall class 1 = 0.5136542320251465
AUC of ROC = 0.7750246111104193
AUC of PRC = 0.6866349101668249
min(+P, Se) = 0.6442154467226177
confusion matrix:
[[2832  422]
 [1045 1057]]
accuracy = 0.7261015772819519
precision class 0 = 0.7304617166519165
precision class 1 = 0.7146720886230469
recall class 0 = 0.8703134655952454
recall class 1 = 0.5028544068336487
AUC of ROC = 0.7624535300767205
AUC of PRC = 0.6746714589723846
min(+P, Se) = 0.645099

In [ ]:
#Runtime aprox 1 hour.

In [ ]:
#Obs. Due to Deprecated warning:
#we have changed the original benchmark version: mimic3models/los3days/logistic/main.py --> sklearn.preprocessing import Imputer, StandardScaler to 
#from sklearn.impute import SimpleImputer 
#and
#imputer = SimpleImputer(missing_values=np.nan, strategy='mean') #insted of 
#from sklearn.preprocessing import Imputer
#imputer = Imputer(missing_values='NaN', strategy='mean', axis=0)

Evaluation:

In [ ]:
!python -m mimic3benchmark.evaluation.evaluate_los_customized /content/gdrive/MyDrive/MIMIC_benchmark/mimic3models/los3days/logistic/predictions/all.all.l2.C0.001.csv --test_listfile /content/los3days/test/listfile.csv --save_file los3days_results_logistic.json

Saving the results in los3days_results_logistic.json ...
{'n_iters': 10000, 'AUC of ROC': {'value': 0.7655737234132598, 'mean': 0.7656816839109378, 'median': 0.7657102575548695, 'std': 0.00679378985525277, '2.5% percentile': 0.7523271382810609, '97.5% percentile': 0.7789856914618944}, 'AUC of PRC': {'value': 0.6657014661218782, 'mean': 0.6662427290011623, 'median': 0.6662546699692931, 'std': 0.011886161832179555, '2.5% percentile': 0.6430259224989597, '97.5% percentile': 0.6896242051503305}, 'min(+P, Se)': {'value': 0.6341821743388835, 'mean': 0.6345678611355393, 'median': 0.6346433770014556, 'std': 0.008878271907149722, '2.5% percentile': 0.6170210088982643, '97.5% percentile': 0.6520259133680957}}


## Bidirectional Standard LSTM:

In [ ]:
!python -um mimic3models.los3days.main --network mimic3models/keras_models/lstm.py --data /content/los3days/ --dim 16 --epochs 50 --timestep 1.0 --depth 2 --dropout 0.3 --mode train --batch_size 8 --output_dir mimic3models/los3days --verbose 1


/content/gdrive/MyDrive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=2, dim=16, dropout=0.3, epochs=50, imputation='previous', l1=0, l2=0, load_state='', lr=0.001, mode='train', network='mimic3models/keras_models/lstm.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=1)
==> using model mimic3models/keras_models/lstm.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_state', 'lr', 'mode', 'network', 'normalizer_state', 'optimizer', 'output_dir', 'prefix', 'save_every', 'size_coef', 'small_part', 'target_repl_coef',

In [ ]:
#runtime: 21:00 to 22:45: aprox. 1.45 horas para 32 epochs

Load epoch 32 y seguir entrenando hasta epoch 35:

In [ ]:
##load previous models states and continue training:
!python -um mimic3models.los3days.main --network mimic3models/keras_models/lstm.py --data /content/los3days/ --load_state mimic3models/los3days/keras_states/k_lstm.n16.d0.3.dep2.bs8.ts1.0.epoch32.test0.5329250078578776.state --dim 16 --epochs 3 --timestep 1.0 --depth 2 --dropout 0.3 --mode train --batch_size 8 --output_dir mimic3models/los3days --verbose 1


/content/gdrive/MyDrive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=2, dim=16, dropout=0.3, epochs=3, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_lstm.n16.d0.3.dep2.bs8.ts1.0.epoch32.test0.5329250078578776.state', lr=0.001, mode='train', network='mimic3models/keras_models/lstm.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=1)
==> using model mimic3models/keras_models/lstm.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_state', 'lr', 'mode', 'network', 'normalizer_stat

In [ ]:
#Escoger epoch con mejor aucpr en validation set
import pandas as pd
aucpr_lstm = pd.read_csv('mimic3models/los3days/keras_logs/k_lstm.n16.d0.3.dep2.bs8.ts1.0.csv', delimiter = ";")
aucpr_lstm.head()

epoch      loss  train_acc  ...  val_prec1  val_rec0  val_rec1
0      0  0.587417   0.733075  ...   0.706294  0.858021  0.528544
1      1  0.567579   0.737321  ...   0.731608  0.878918  0.510942
2      2  0.562441   0.739853  ...   0.734007  0.878611  0.518554
3      3  0.557922   0.741772  ...   0.721559  0.863860  0.546146
4      4  0.553568   0.741935  ...   0.736176  0.878304  0.525690

[5 rows x 19 columns]

In [ ]:
aucpr_lstm['epoch'][aucpr_lstm['val_auprc'].idxmax()]+1  #obs. epoch en dataframe empieza por 0, por lo tanto es epoch 35


35

In [ ]:
aucpr_lstm['val_auprc'].max()

0.723691527840633

Predictiong in Test

In [ ]:
#revisando Keras_logs: epoch 35 tiene mejor performance (en validation dataset) en aucpr 0.7236 

In [ ]:
!python3 -um mimic3models.los3days.main --network mimic3models/keras_models/lstm.py --data /content/los3days/ --dim 16 --timestep 1.0 --depth 2 --dropout 0.3 --batch_size 8 --epochs 50 --load_state mimic3models/los3days/keras_states/k_lstm.n16.d0.3.dep2.bs8.ts1.0.epoch35.test0.5336932081442612.state --output_dir mimic3models/los3days --mode test

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=2, dim=16, dropout=0.3, epochs=50, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_lstm.n16.d0.3.dep2.bs8.ts1.0.epoch35.test0.5336932081442612.state', lr=0.001, mode='test', network='mimic3models/keras_models/lstm.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=2)
==> using model mimic3models/keras_models/lstm.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_state', 'lr', 'mode', 'network', 'normalizer_sta

Evaluation

In [ ]:
!python3 -m mimic3benchmark.evaluation.evaluate_los_customized /content/gdrive/MyDrive/MIMIC_benchmark/mimic3models/los3days/test_predictions/k_lstm.n16.d0.3.dep2.bs8.ts1.0.epoch35.test0.5336932081442612.state.csv --test_listfile /content/los3days/test/listfile.csv --save_file los3days_results_lstm_epoch35.json

Saving the results in los3days_results_lstm_epoch35.json ...
{'n_iters': 10000, 'AUC of ROC': {'value': 0.7853261296719508, 'mean': 0.7853141394894224, 'median': 0.7853430574482496, 'std': 0.006560660529273359, '2.5% percentile': 0.7723216687337059, '97.5% percentile': 0.7980868993639059}, 'AUC of PRC': {'value': 0.7126836374588981, 'mean': 0.7127860644603167, 'median': 0.7129193218321926, 'std': 0.010634197975967487, '2.5% percentile': 0.6917056609329736, '97.5% percentile': 0.7332218240290677}, 'min(+P, Se)': {'value': 0.6518119490695397, 'mean': 0.6514553968905032, 'median': 0.6515301085883515, 'std': 0.008640472447633362, '2.5% percentile': 0.6344030802481153, '97.5% percentile': 0.6680974876750959}}


## Channel-wise LSTM:

In [ ]:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_lstms.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --epochs 6 --mode train --size_coef 4.0 --output_dir mimic3models/los3days

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=6, imputation='previous', l1=0, l2=0, load_state='', lr=0.001, mode='train', network='mimic3models/keras_models/channel_wise_lstms.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=2)
==> using model mimic3models/keras_models/channel_wise_lstms.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_state', 'lr', 'mode', 'network', 'normalizer_state', 'optimizer', 'output_dir', 'prefix', 'save_every', 'small_part', 'targ

In [ ]:
#load previous models states and continue training:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_lstms.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --epochs 3 --load_state mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch7.test0.535226437607481.state  --mode train --size_coef 4.0 --output_dir mimic3models/los3days

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=3, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch7.test0.535226437607481.state', lr=0.001, mode='train', network='mimic3models/keras_models/channel_wise_lstms.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=2)
==> using model mimic3models/keras_models/channel_wise_lstms.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_sta

Load epoch 10 and continue training:


In [ ]:
#load previous models states and continue training:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_lstms.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --epochs 5 --load_state mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch10.test0.5291586841906604.state  --mode train --size_coef 4.0 --output_dir mimic3models/los3days

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=5, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch10.test0.5291586841906604.state', lr=0.001, mode='train', network='mimic3models/keras_models/channel_wise_lstms.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=2)
==> using model mimic3models/keras_models/channel_wise_lstms.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_s

Load epoch 15 and continue training:


In [ ]:
#load previous models states and continue training:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_lstms.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --epochs 5 --load_state mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch15.test0.5296070242524592.state  --mode train --size_coef 4.0 --output_dir mimic3models/los3days --verbose 1

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=5, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch15.test0.5296070242524592.state', lr=0.001, mode='train', network='mimic3models/keras_models/channel_wise_lstms.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=1)
==> using model mimic3models/keras_models/channel_wise_lstms.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_s

Load epoch 20 and continue training:

In [ ]:
#load previous models states and continue training:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_lstms.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --epochs 5 --load_state mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch20.test0.5294502408002365.state  --mode train --size_coef 4.0 --output_dir mimic3models/los3days --verbose 1

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=5, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch20.test0.5294502408002365.state', lr=0.001, mode='train', network='mimic3models/keras_models/channel_wise_lstms.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=1)
==> using model mimic3models/keras_models/channel_wise_lstms.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_s

Load epoch 25 and continue training:

In [ ]:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_lstms.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --epochs 5 --load_state mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch25.test0.527318338300863.state  --mode train --size_coef 4.0 --output_dir mimic3models/los3days --verbose 1

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=5, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch25.test0.527318338300863.state', lr=0.001, mode='train', network='mimic3models/keras_models/channel_wise_lstms.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=1)
==> using model mimic3models/keras_models/channel_wise_lstms.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_st

Load epoch 30 and continue training:

In [ ]:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_lstms.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --epochs 2 --load_state mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch30.test0.5253430250155738.state  --mode train --size_coef 4.0 --output_dir mimic3models/los3days --verbose 1

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=2, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch30.test0.5253430250155738.state', lr=0.001, mode='train', network='mimic3models/keras_models/channel_wise_lstms.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=1)
==> using model mimic3models/keras_models/channel_wise_lstms.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_s

Load epoch 32 and continue training:

In [ ]:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_lstms.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --epochs 3 --load_state mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch32.test0.527357709372195.state  --mode train --size_coef 4.0 --output_dir mimic3models/los3days --verbose 1

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=3, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch32.test0.527357709372195.state', lr=0.001, mode='train', network='mimic3models/keras_models/channel_wise_lstms.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=1)
==> using model mimic3models/keras_models/channel_wise_lstms.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_st

In [ ]:
#Escoger epoch con mejor aucpr en validation set
import pandas as pd
aucpr_channelwise_lstm = pd.read_csv('mimic3models/los3days/keras_logs/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.csv', delimiter = ";")
aucpr_channelwise_lstm.head()

epoch      loss  train_acc  ...  val_prec1  val_rec0  val_rec1
0      0  0.584072   0.735361  ...   0.729278  0.875538  0.519030
1      1  0.567249   0.740588  ...   0.712969  0.856484  0.551855
2      2  0.557812   0.741445  ...   0.731266  0.872157  0.538535
3      3  0.552697   0.744508  ...   0.736773  0.876152  0.536632
4      4  0.550841   0.746100  ...   0.727875  0.861094  0.575167

[5 rows x 19 columns]

In [ ]:
aucpr_channelwise_lstm['epoch'][aucpr_channelwise_lstm['val_auprc'].idxmax()]+1  #obs. epoch en dataframe empieza por 0


28

In [ ]:
aucpr_channelwise_lstm['val_auprc'].max()

0.7367447704458944

Predicting on Test with epoch 28 (la mejor en aucpr sobre validation set, aucpr: 0.736):

In [ ]:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_lstms.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --epochs 5 --load_state mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch28.test0.5260214427676639.state --mode test --size_coef 4.0 --output_dir mimic3models/los3days

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=5, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch28.test0.5260214427676639.state', lr=0.001, mode='test', network='mimic3models/keras_models/channel_wise_lstms.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=2)
==> using model mimic3models/keras_models/channel_wise_lstms.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_st

Evaluation with epoch 28:

In [23]:
!python -m mimic3benchmark.evaluation.evaluate_los_customized /content/gdrive/MyDrive/MIMIC_benchmark/mimic3models/los3days/test_predictions/k_channel_wise_lstms.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch28.test0.5260214427676639.state.csv --test_listfile /content/los3days/test/listfile.csv --save_file los3days_results_channelWise_lstm_epoch28.json

Saving the results in los3days_results_channelWise_lstm_epoch28.json ...
{'n_iters': 10000, 'AUC of ROC': {'value': 0.7903547115512509, 'mean': 0.790298180684303, 'median': 0.7902933565476594, 'std': 0.006519500274398229, '2.5% percentile': 0.7776368054072657, '97.5% percentile': 0.803104889051896}, 'AUC of PRC': {'value': 0.7193089654741814, 'mean': 0.7193263311805393, 'median': 0.7194798695816172, 'std': 0.010782362155847441, '2.5% percentile': 0.6978825156489088, '97.5% percentile': 0.7399949245685141}, 'min(+P, Se)': {'value': 0.6557296767874633, 'mean': 0.655416454879351, 'median': 0.6555744673620296, 'std': 0.008785065384158082, '2.5% percentile': 0.6377473363774734, '97.5% percentile': 0.6722734356994602}}


## Channel wise GRU

In [ ]:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_grus.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --epochs 7 --mode train --size_coef 4.0 --output_dir mimic3models/los3days

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=7, imputation='previous', l1=0, l2=0, load_state='', lr=0.001, mode='train', network='mimic3models/keras_models/channel_wise_grus.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=2)
==> using model mimic3models/keras_models/channel_wise_grus.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_state', 'lr', 'mode', 'network', 'normalizer_state', 'optimizer', 'output_dir', 'prefix', 'save_every', 'small_part', 'target

In [ ]:
#Total 2h19min para 7 epochs

Load epoch 7 and continue training:

In [ ]:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_grus.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --epochs 8 --load_state mimic3models/los3days/keras_states/k_channel_wise_grus.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch7.test0.5370793866746189.state  --mode train --size_coef 4.0 --output_dir mimic3models/los3days

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=8, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_channel_wise_grus.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch7.test0.5370793866746189.state', lr=0.001, mode='train', network='mimic3models/keras_models/channel_wise_grus.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=2)
==> using model mimic3models/keras_models/channel_wise_grus.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_state

In [ ]:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_grus.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --epochs 5 --load_state mimic3models/los3days/keras_states/k_channel_wise_grus.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch15.test0.5309026377071099.state  --mode train --size_coef 4.0 --output_dir mimic3models/los3days

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=5, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_channel_wise_grus.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch15.test0.5309026377071099.state', lr=0.001, mode='train', network='mimic3models/keras_models/channel_wise_grus.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=2)
==> using model mimic3models/keras_models/channel_wise_grus.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_stat

In [ ]:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_grus.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --epochs 5 --load_state mimic3models/los3days/keras_states/k_channel_wise_grus.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch20.test0.529586777858898.state  --mode train --size_coef 4.0 --output_dir mimic3models/los3days

/content/gdrive/MyDrive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=5, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_channel_wise_grus.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch20.test0.529586777858898.state', lr=0.001, mode='train', network='mimic3models/keras_models/channel_wise_grus.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=2)
==> using model mimic3models/keras_models/channel_wise_grus.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_state'

In [ ]:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_grus.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --epochs 5 --load_state mimic3models/los3days/keras_states/k_channel_wise_grus.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch25.test0.5310363372463713.state  --mode train --size_coef 4.0 --output_dir mimic3models/los3days

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=5, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_channel_wise_grus.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch25.test0.5310363372463713.state', lr=0.001, mode='train', network='mimic3models/keras_models/channel_wise_grus.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=2)
==> using model mimic3models/keras_models/channel_wise_grus.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_stat

In [ ]:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_grus.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --epochs 5 --load_state mimic3models/los3days/keras_states/k_channel_wise_grus.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch30.test0.5312995400510678.state  --mode train --size_coef 4.0 --output_dir mimic3models/los3days

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=5, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_channel_wise_grus.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch30.test0.5312995400510678.state', lr=0.001, mode='train', network='mimic3models/keras_models/channel_wise_grus.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=2)
==> using model mimic3models/keras_models/channel_wise_grus.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_stat

In [ ]:
#Escoger epoch con mejor aucpr en validation set
import pandas as pd
aucpr_channelwise_gru = pd.read_csv('mimic3models/los3days/keras_logs/k_channel_wise_grus.n8.szc4.0.d0.3.dep1.bs8.ts1.0.csv', delimiter = ";")
aucpr_channelwise_gru.head()

epoch      loss  train_acc  ...  val_prec1  val_rec0  val_rec1
0      0  0.591901   0.737035  ...   0.721616  0.866626  0.535205
1      1  0.566669   0.737240  ...   0.738482  0.881377  0.518554
2      2  0.559690   0.742425  ...   0.743448  0.885679  0.512845
3      3  0.554861   0.744835  ...   0.733291  0.871235  0.548049
4      4  0.553744   0.742058  ...   0.715224  0.853411  0.569933

[5 rows x 19 columns]

In [ ]:
aucpr_channelwise_gru['epoch'][aucpr_channelwise_gru['val_auprc'].idxmax()]+1  #obs. epoch en dataframe empieza por 0


20

In [ ]:
aucpr_channelwise_gru['val_auprc'].max()

0.7272804044949749


Predicting on Test with epoch 20 (la mejor en aucpr sobre validation set, aucpr: 0.727):

In [ ]:
!python -um mimic3models.los3days.main --data /content/los3days/ --network mimic3models/keras_models/channel_wise_grus.py --dim 8 --depth 1 --batch_size 8 --dropout 0.3 --timestep 1.0 --load_state mimic3models/los3days/keras_states/k_channel_wise_grus.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch20.test0.529586777858898.state --mode test --size_coef 4.0 --output_dir mimic3models/los3days

/content/gdrive/My Drive/MIMIC_benchmark/mimic3models/los3days/main.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using TensorFlow backend.
Namespace(batch_norm=False, batch_size=8, beta_1=0.9, data='/content/los3days/', depth=1, dim=8, dropout=0.3, epochs=100, imputation='previous', l1=0, l2=0, load_state='mimic3models/los3days/keras_states/k_channel_wise_grus.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch20.test0.529586777858898.state', lr=0.001, mode='test', network='mimic3models/keras_models/channel_wise_grus.py', normalizer_state=None, optimizer='adam', output_dir='mimic3models/los3days', prefix='', rec_dropout=0.0, save_every=1, size_coef=4.0, small_part=False, target_repl_coef=0.0, timestep=1.0, verbose=2)
==> using model mimic3models/keras_models/channel_wise_grus.py
==> not used params in network class: dict_keys(['batch_size', 'beta_1', 'data', 'epochs', 'imputation', 'l1', 'l2', 'load_stat

95% Confidence Interval evaluation with epoch 20:

In [24]:
!python -m mimic3benchmark.evaluation.evaluate_los_customized /content/gdrive/MyDrive/MIMIC_benchmark/mimic3models/los3days/test_predictions/k_channel_wise_grus.n8.szc4.0.d0.3.dep1.bs8.ts1.0.epoch20.test0.529586777858898.state.csv --test_listfile /content/los3days/test/listfile.csv --save_file los3days_results_channelWise_gru_epoch20.json

Saving the results in los3days_results_channelWise_gru_epoch20.json ...
{'n_iters': 10000, 'AUC of ROC': {'value': 0.7891676793509147, 'mean': 0.7892349048064301, 'median': 0.7893228309218384, 'std': 0.006504602507796131, '2.5% percentile': 0.7763366481719772, '97.5% percentile': 0.8018266199334211}, 'AUC of PRC': {'value': 0.7166283490326372, 'mean': 0.7168859592473115, 'median': 0.7169850771954692, 'std': 0.010578617115684972, '2.5% percentile': 0.695805741324334, '97.5% percentile': 0.7372454640528683}, 'min(+P, Se)': {'value': 0.6576885406464251, 'mean': 0.6584358388409657, 'median': 0.658477250240256, 'std': 0.008654647055178, '2.5% percentile': 0.6414264777723497, '97.5% percentile': 0.6750837206155238}}
